{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "import h5py\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "# from os import listdir, path"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## GCTX File"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "gctx = h5py.File('GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gctx', 'r')"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "gctx['0/META/COL/id'].shape, gctx['0/META/ROW/id'].shape, gctx['0/DATA/0/matrix'].shape"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "((17382,), (56200,), (17382, 56200))"
      ]
     },
     "metadata": {},
     "execution_count": 3
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "gctx['0/META/COL/id'][:5]"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "array([b'GTEX-1117F-0226-SM-5GZZ7', b'GTEX-1117F-0426-SM-5EGHI',\n",
       "       b'GTEX-1117F-0526-SM-5EGHJ', b'GTEX-1117F-0626-SM-5N9CS',\n",
       "       b'GTEX-1117F-0726-SM-5GIEN'], dtype='|S32')"
      ]
     },
     "metadata": {},
     "execution_count": 4
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "gctx['0/META/ROW/id'][:5]"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "array([b'ENSG00000223972.5', b'ENSG00000227232.5', b'ENSG00000278267.1',\n",
       "       b'ENSG00000243485.5', b'ENSG00000237613.2'], dtype='|S27')"
      ]
     },
     "metadata": {},
     "execution_count": 5
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Divide samples by tissue"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "source": [
    "sample_meta = pd.read_csv('GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt', sep='\\t')\n",
    "sample_meta['sub'] = sample_meta['SAMPID'].apply(lambda x: '-'.join(x.split('-')[:2]))\n",
    "sample_meta = sample_meta[['SAMPID', 'SMTS', 'sub']].set_index('SAMPID')\n",
    "sample_meta.head()"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "                                SMTS         sub\n",
       "SAMPID                                          \n",
       "GTEX-1117F-0003-SM-58Q7G       Blood  GTEX-1117F\n",
       "GTEX-1117F-0003-SM-5DWSB       Blood  GTEX-1117F\n",
       "GTEX-1117F-0003-SM-6WBT7       Blood  GTEX-1117F\n",
       "GTEX-1117F-0011-R10a-SM-AHZ7F  Brain  GTEX-1117F\n",
       "GTEX-1117F-0011-R10b-SM-CYKQ8  Brain  GTEX-1117F"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SMTS</th>\n",
       "      <th>sub</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0003-SM-58Q7G</th>\n",
       "      <td>Blood</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0003-SM-5DWSB</th>\n",
       "      <td>Blood</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0003-SM-6WBT7</th>\n",
       "      <td>Blood</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0011-R10a-SM-AHZ7F</th>\n",
       "      <td>Brain</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0011-R10b-SM-CYKQ8</th>\n",
       "      <td>Brain</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "execution_count": 6
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Divide subjects by age"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "source": [
    "meta = pd.read_csv('GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt', sep='\\t')\n",
    "meta = meta.set_index('SUBJID')\n",
    "\n",
    "sample_meta['age'] = [meta.loc[row.sub, 'AGE'] for row in sample_meta.itertuples()]\n",
    "sample_meta['sex'] = [meta.loc[row.sub, 'SEX'] for row in sample_meta.itertuples()]\n",
    "sample_meta.head()"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "                                SMTS         sub    age  sex\n",
       "SAMPID                                                      \n",
       "GTEX-1117F-0003-SM-58Q7G       Blood  GTEX-1117F  60-69    2\n",
       "GTEX-1117F-0003-SM-5DWSB       Blood  GTEX-1117F  60-69    2\n",
       "GTEX-1117F-0003-SM-6WBT7       Blood  GTEX-1117F  60-69    2\n",
       "GTEX-1117F-0011-R10a-SM-AHZ7F  Brain  GTEX-1117F  60-69    2\n",
       "GTEX-1117F-0011-R10b-SM-CYKQ8  Brain  GTEX-1117F  60-69    2"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SMTS</th>\n",
       "      <th>sub</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0003-SM-58Q7G</th>\n",
       "      <td>Blood</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "      <td>60-69</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0003-SM-5DWSB</th>\n",
       "      <td>Blood</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "      <td>60-69</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0003-SM-6WBT7</th>\n",
       "      <td>Blood</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "      <td>60-69</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0011-R10a-SM-AHZ7F</th>\n",
       "      <td>Brain</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "      <td>60-69</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0011-R10b-SM-CYKQ8</th>\n",
       "      <td>Brain</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "      <td>60-69</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "execution_count": 7
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "source": [
    "sample_meta.shape[0]"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "22951"
      ]
     },
     "metadata": {},
     "execution_count": 9
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "source": [
    "# get rid of discrepancies between metadata and data\n",
    "sample_meta = sample_meta.loc[gctx['0/META/COL/id'].asstr()[:]]\n",
    "sample_meta.shape"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "(17382, 4)"
      ]
     },
     "metadata": {},
     "execution_count": 11
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Map Ensembl genes"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "source": [
    "# map ensembl genes\n",
    "gene_info = pd.read_csv('Homo_sapiens.gene_info', sep='\\t')[['Symbol', 'dbXrefs']]\n",
    "gene_info['ensembl'] = [row.dbXrefs.split('Ensembl:')[1] if 'Ensembl' in row.dbXrefs else np.nan for row in gene_info.itertuples()]\n",
    "gene_info = gene_info[['Symbol', 'ensembl']].dropna(subset=['ensembl']).set_index('ensembl')\n",
    "gene_info.shape"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "(35161, 1)"
      ]
     },
     "metadata": {},
     "execution_count": 16
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "source": [
    "gene_info.head()"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "                Symbol\n",
       "ensembl               \n",
       "ENSG00000121410   A1BG\n",
       "ENSG00000175899    A2M\n",
       "ENSG00000256069  A2MP1\n",
       "ENSG00000171428   NAT1\n",
       "ENSG00000156006   NAT2"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000121410</th>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000175899</th>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000256069</th>\n",
       "      <td>A2MP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000171428</th>\n",
       "      <td>NAT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000156006</th>\n",
       "      <td>NAT2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "execution_count": 17
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "source": [
    "to_keep = []\n",
    "for g in gctx['0/META/ROW/id'][()].astype(str):\n",
    "    if g.split('.')[0] in gene_info.index:\n",
    "        to_keep.append(g)\n",
    "\n",
    "len(to_keep)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "32024"
      ]
     },
     "metadata": {},
     "execution_count": 19
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "source": [
    "def slice_matrix(gctx, rids): \n",
    "    all_rids = gctx['/0/META/ROW/id'].asstr()[:]\n",
    "    r_mask = np.in1d(all_rids, rids)\n",
    "    rids_subset = all_rids[r_mask].tolist()\n",
    "    r_indices = np.array([rids_subset.index(id_) for id_ in rids])\n",
    "    \n",
    "    mat = gctx['/0/DATA/0/matrix'][()]\n",
    "    submat = mat[:, r_mask][:, r_indices]\n",
    "\n",
    "    return submat"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "source": [
    "row_ids = sorted(to_keep)\n",
    "pruned_data = slice_matrix(gctx, row_ids)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "source": [
    "pruned_data.shape"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "(17382, 32024)"
      ]
     },
     "metadata": {},
     "execution_count": 22
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "source": [
    "data_df = pd.DataFrame(data=pruned_data, columns=row_ids, index=gctx['0/META/COL/id'].asstr()[:])\n",
    "data_df.head()"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "                          ENSG00000000003.14  ENSG00000000005.5  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7              2257.0              434.0   \n",
       "GTEX-1117F-0426-SM-5EGHI               408.0               24.0   \n",
       "GTEX-1117F-0526-SM-5EGHJ              1082.0               74.0   \n",
       "GTEX-1117F-0626-SM-5N9CS              2822.0              245.0   \n",
       "GTEX-1117F-0726-SM-5GIEN               778.0                2.0   \n",
       "\n",
       "                          ENSG00000000419.12  ENSG00000000457.13  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7              1231.0               688.0   \n",
       "GTEX-1117F-0426-SM-5EGHI              2447.0               222.0   \n",
       "GTEX-1117F-0526-SM-5EGHJ              1052.0               447.0   \n",
       "GTEX-1117F-0626-SM-5N9CS              1347.0               766.0   \n",
       "GTEX-1117F-0726-SM-5GIEN               905.0               335.0   \n",
       "\n",
       "                          ENSG00000000460.16  ENSG00000000938.12  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7               174.0              2175.0   \n",
       "GTEX-1117F-0426-SM-5EGHI               112.0               216.0   \n",
       "GTEX-1117F-0526-SM-5EGHJ               118.0               933.0   \n",
       "GTEX-1117F-0626-SM-5N9CS               214.0              1011.0   \n",
       "GTEX-1117F-0726-SM-5GIEN               113.0               388.0   \n",
       "\n",
       "                          ENSG00000000971.15  ENSG00000001036.13  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7             21961.0              3438.0   \n",
       "GTEX-1117F-0426-SM-5EGHI               834.0               281.0   \n",
       "GTEX-1117F-0526-SM-5EGHJ             14106.0               758.0   \n",
       "GTEX-1117F-0626-SM-5N9CS             39533.0              1712.0   \n",
       "GTEX-1117F-0726-SM-5GIEN             62151.0              1245.0   \n",
       "\n",
       "                          ENSG00000001084.10  ENSG00000001167.14  ...  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7              1204.0              1296.0  ...   \n",
       "GTEX-1117F-0426-SM-5EGHI               414.0               632.0  ...   \n",
       "GTEX-1117F-0526-SM-5EGHJ               648.0               954.0  ...   \n",
       "GTEX-1117F-0626-SM-5N9CS              1091.0              1473.0  ...   \n",
       "GTEX-1117F-0726-SM-5GIEN               521.0              1085.0  ...   \n",
       "\n",
       "                          ENSG00000284328.1  ENSG00000284373.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7                0.0                0.0   \n",
       "GTEX-1117F-0426-SM-5EGHI                0.0                0.0   \n",
       "GTEX-1117F-0526-SM-5EGHJ                0.0                0.0   \n",
       "GTEX-1117F-0626-SM-5N9CS                0.0                0.0   \n",
       "GTEX-1117F-0726-SM-5GIEN                0.0                0.0   \n",
       "\n",
       "                          ENSG00000284377.1  ENSG00000284391.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7                0.0                1.0   \n",
       "GTEX-1117F-0426-SM-5EGHI                0.0               70.0   \n",
       "GTEX-1117F-0526-SM-5EGHJ                0.0               55.0   \n",
       "GTEX-1117F-0626-SM-5N9CS                0.0               31.0   \n",
       "GTEX-1117F-0726-SM-5GIEN                0.0               55.0   \n",
       "\n",
       "                          ENSG00000284395.1  ENSG00000284438.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7               11.0                0.0   \n",
       "GTEX-1117F-0426-SM-5EGHI                6.0                0.0   \n",
       "GTEX-1117F-0526-SM-5EGHJ                7.0                0.0   \n",
       "GTEX-1117F-0626-SM-5N9CS                0.0                0.0   \n",
       "GTEX-1117F-0726-SM-5GIEN                0.0                0.0   \n",
       "\n",
       "                          ENSG00000284471.1  ENSG00000284516.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7               13.0                0.0   \n",
       "GTEX-1117F-0426-SM-5EGHI                4.0                0.0   \n",
       "GTEX-1117F-0526-SM-5EGHJ               50.0                0.0   \n",
       "GTEX-1117F-0626-SM-5N9CS               12.0                0.0   \n",
       "GTEX-1117F-0726-SM-5GIEN               10.0                0.0   \n",
       "\n",
       "                          ENSG00000284523.1  ENSG00000284546.1  \n",
       "GTEX-1117F-0226-SM-5GZZ7               24.0                0.0  \n",
       "GTEX-1117F-0426-SM-5EGHI               15.0                0.0  \n",
       "GTEX-1117F-0526-SM-5EGHJ               24.0                0.0  \n",
       "GTEX-1117F-0626-SM-5N9CS               14.0                0.0  \n",
       "GTEX-1117F-0726-SM-5GIEN                8.0                0.0  \n",
       "\n",
       "[5 rows x 32024 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "      <th>ENSG00000000938.12</th>\n",
       "      <th>ENSG00000000971.15</th>\n",
       "      <th>ENSG00000001036.13</th>\n",
       "      <th>ENSG00000001084.10</th>\n",
       "      <th>ENSG00000001167.14</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000284328.1</th>\n",
       "      <th>ENSG00000284373.1</th>\n",
       "      <th>ENSG00000284377.1</th>\n",
       "      <th>ENSG00000284391.1</th>\n",
       "      <th>ENSG00000284395.1</th>\n",
       "      <th>ENSG00000284438.1</th>\n",
       "      <th>ENSG00000284471.1</th>\n",
       "      <th>ENSG00000284516.1</th>\n",
       "      <th>ENSG00000284523.1</th>\n",
       "      <th>ENSG00000284546.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>2257.0</td>\n",
       "      <td>434.0</td>\n",
       "      <td>1231.0</td>\n",
       "      <td>688.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>2175.0</td>\n",
       "      <td>21961.0</td>\n",
       "      <td>3438.0</td>\n",
       "      <td>1204.0</td>\n",
       "      <td>1296.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <td>408.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>2447.0</td>\n",
       "      <td>222.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>216.0</td>\n",
       "      <td>834.0</td>\n",
       "      <td>281.0</td>\n",
       "      <td>414.0</td>\n",
       "      <td>632.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <td>1082.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>1052.0</td>\n",
       "      <td>447.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>933.0</td>\n",
       "      <td>14106.0</td>\n",
       "      <td>758.0</td>\n",
       "      <td>648.0</td>\n",
       "      <td>954.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <td>2822.0</td>\n",
       "      <td>245.0</td>\n",
       "      <td>1347.0</td>\n",
       "      <td>766.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>1011.0</td>\n",
       "      <td>39533.0</td>\n",
       "      <td>1712.0</td>\n",
       "      <td>1091.0</td>\n",
       "      <td>1473.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <td>778.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>905.0</td>\n",
       "      <td>335.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>388.0</td>\n",
       "      <td>62151.0</td>\n",
       "      <td>1245.0</td>\n",
       "      <td>521.0</td>\n",
       "      <td>1085.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32024 columns</p>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "execution_count": 32
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Remove duplicates by variance"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "source": [
    "var_df = data_df.var(axis=0).to_frame(name='Var')\n",
    "var_df['Ens'] = var_df.index.map(lambda x: x.split('.')[0])\n",
    "var_df.head()"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "                             Var              Ens\n",
       "ENSG00000000003.14  4.561226e+06  ENSG00000000003\n",
       "ENSG00000000005.5   2.789238e+05  ENSG00000000005\n",
       "ENSG00000000419.12  5.480776e+05  ENSG00000000419\n",
       "ENSG00000000457.13  1.864860e+05  ENSG00000000457\n",
       "ENSG00000000460.16  8.066638e+04  ENSG00000000460"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Var</th>\n",
       "      <th>Ens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <td>4.561226e+06</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <td>2.789238e+05</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <td>5.480776e+05</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <td>1.864860e+05</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "      <td>8.066638e+04</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "execution_count": 37
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "source": [
    "keep = var_df.sort_values(by=['Ens', 'Var'], ascending=True).drop_duplicates(subset=['Ens'], keep='last').index"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "source": [
    "full_data_df = data_df[keep].T"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Map gene symbols"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "source": [
    "gene_info = gene_info.loc[gene_info.index.drop_duplicates(keep='first')]"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "source": [
    "full_data_df.index = full_data_df.index.map(lambda x: gene_info.loc[x.split('.')[0], 'Symbol'])"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "source": [
    "full_data_df.head()"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "          GTEX-1117F-0226-SM-5GZZ7  GTEX-1117F-0426-SM-5EGHI  \\\n",
       "TSPAN6                      2257.0                     408.0   \n",
       "TNMD                         434.0                      24.0   \n",
       "DPM1                        1231.0                    2447.0   \n",
       "SCYL3                        688.0                     222.0   \n",
       "C1orf112                     174.0                     112.0   \n",
       "\n",
       "          GTEX-1117F-0526-SM-5EGHJ  GTEX-1117F-0626-SM-5N9CS  \\\n",
       "TSPAN6                      1082.0                    2822.0   \n",
       "TNMD                          74.0                     245.0   \n",
       "DPM1                        1052.0                    1347.0   \n",
       "SCYL3                        447.0                     766.0   \n",
       "C1orf112                     118.0                     214.0   \n",
       "\n",
       "          GTEX-1117F-0726-SM-5GIEN  GTEX-1117F-1326-SM-5EGHH  \\\n",
       "TSPAN6                       778.0                    3417.0   \n",
       "TNMD                           2.0                     625.0   \n",
       "DPM1                         905.0                     820.0   \n",
       "SCYL3                        335.0                     295.0   \n",
       "C1orf112                     113.0                     117.0   \n",
       "\n",
       "          GTEX-1117F-2426-SM-5EGGH  GTEX-1117F-2526-SM-5GZY6  \\\n",
       "TSPAN6                      2507.0                    2167.0   \n",
       "TNMD                          13.0                      23.0   \n",
       "DPM1                        1041.0                    1615.0   \n",
       "SCYL3                        635.0                    1348.0   \n",
       "C1orf112                     282.0                     312.0   \n",
       "\n",
       "          GTEX-1117F-2826-SM-5GZXL  GTEX-1117F-2926-SM-5GZYI  ...  \\\n",
       "TSPAN6                      4622.0                     852.0  ...   \n",
       "TNMD                        1837.0                      60.0  ...   \n",
       "DPM1                        1269.0                    1313.0  ...   \n",
       "SCYL3                        871.0                     697.0  ...   \n",
       "C1orf112                     303.0                     205.0  ...   \n",
       "\n",
       "          GTEX-ZZPU-1126-SM-5N9CW  GTEX-ZZPU-1226-SM-5N9CK  \\\n",
       "TSPAN6                      926.0                   2279.0   \n",
       "TNMD                         22.0                      0.0   \n",
       "DPM1                       1583.0                   2101.0   \n",
       "SCYL3                       675.0                   1252.0   \n",
       "C1orf112                    146.0                    279.0   \n",
       "\n",
       "          GTEX-ZZPU-1326-SM-5GZWS  GTEX-ZZPU-1426-SM-5GZZ6  \\\n",
       "TSPAN6                     1474.0                   1142.0   \n",
       "TNMD                          8.0                      6.0   \n",
       "DPM1                       1080.0                   1286.0   \n",
       "SCYL3                      1075.0                    720.0   \n",
       "C1orf112                    262.0                    129.0   \n",
       "\n",
       "          GTEX-ZZPU-1826-SM-5E43L  GTEX-ZZPU-2126-SM-5EGIU  \\\n",
       "TSPAN6                      553.0                   7515.0   \n",
       "TNMD                          5.0                     83.0   \n",
       "DPM1                       1501.0                   1340.0   \n",
       "SCYL3                       784.0                    943.0   \n",
       "C1orf112                    241.0                    323.0   \n",
       "\n",
       "          GTEX-ZZPU-2226-SM-5EGIV  GTEX-ZZPU-2426-SM-5E44I  \\\n",
       "TSPAN6                     2503.0                    574.0   \n",
       "TNMD                          0.0                      4.0   \n",
       "DPM1                       1076.0                   1448.0   \n",
       "SCYL3                       681.0                    787.0   \n",
       "C1orf112                    300.0                    184.0   \n",
       "\n",
       "          GTEX-ZZPU-2626-SM-5E45Y  GTEX-ZZPU-2726-SM-5NQ8O  \n",
       "TSPAN6                      246.0                   3502.0  \n",
       "TNMD                        309.0                   1182.0  \n",
       "DPM1                       1312.0                   1337.0  \n",
       "SCYL3                       983.0                   1235.0  \n",
       "C1orf112                    100.0                    382.0  \n",
       "\n",
       "[5 rows x 17382 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <th>GTEX-1117F-1326-SM-5EGHH</th>\n",
       "      <th>GTEX-1117F-2426-SM-5EGGH</th>\n",
       "      <th>GTEX-1117F-2526-SM-5GZY6</th>\n",
       "      <th>GTEX-1117F-2826-SM-5GZXL</th>\n",
       "      <th>GTEX-1117F-2926-SM-5GZYI</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-ZZPU-1126-SM-5N9CW</th>\n",
       "      <th>GTEX-ZZPU-1226-SM-5N9CK</th>\n",
       "      <th>GTEX-ZZPU-1326-SM-5GZWS</th>\n",
       "      <th>GTEX-ZZPU-1426-SM-5GZZ6</th>\n",
       "      <th>GTEX-ZZPU-1826-SM-5E43L</th>\n",
       "      <th>GTEX-ZZPU-2126-SM-5EGIU</th>\n",
       "      <th>GTEX-ZZPU-2226-SM-5EGIV</th>\n",
       "      <th>GTEX-ZZPU-2426-SM-5E44I</th>\n",
       "      <th>GTEX-ZZPU-2626-SM-5E45Y</th>\n",
       "      <th>GTEX-ZZPU-2726-SM-5NQ8O</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>2257.0</td>\n",
       "      <td>408.0</td>\n",
       "      <td>1082.0</td>\n",
       "      <td>2822.0</td>\n",
       "      <td>778.0</td>\n",
       "      <td>3417.0</td>\n",
       "      <td>2507.0</td>\n",
       "      <td>2167.0</td>\n",
       "      <td>4622.0</td>\n",
       "      <td>852.0</td>\n",
       "      <td>...</td>\n",
       "      <td>926.0</td>\n",
       "      <td>2279.0</td>\n",
       "      <td>1474.0</td>\n",
       "      <td>1142.0</td>\n",
       "      <td>553.0</td>\n",
       "      <td>7515.0</td>\n",
       "      <td>2503.0</td>\n",
       "      <td>574.0</td>\n",
       "      <td>246.0</td>\n",
       "      <td>3502.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>434.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>245.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>625.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1837.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>309.0</td>\n",
       "      <td>1182.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>1231.0</td>\n",
       "      <td>2447.0</td>\n",
       "      <td>1052.0</td>\n",
       "      <td>1347.0</td>\n",
       "      <td>905.0</td>\n",
       "      <td>820.0</td>\n",
       "      <td>1041.0</td>\n",
       "      <td>1615.0</td>\n",
       "      <td>1269.0</td>\n",
       "      <td>1313.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1583.0</td>\n",
       "      <td>2101.0</td>\n",
       "      <td>1080.0</td>\n",
       "      <td>1286.0</td>\n",
       "      <td>1501.0</td>\n",
       "      <td>1340.0</td>\n",
       "      <td>1076.0</td>\n",
       "      <td>1448.0</td>\n",
       "      <td>1312.0</td>\n",
       "      <td>1337.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>688.0</td>\n",
       "      <td>222.0</td>\n",
       "      <td>447.0</td>\n",
       "      <td>766.0</td>\n",
       "      <td>335.0</td>\n",
       "      <td>295.0</td>\n",
       "      <td>635.0</td>\n",
       "      <td>1348.0</td>\n",
       "      <td>871.0</td>\n",
       "      <td>697.0</td>\n",
       "      <td>...</td>\n",
       "      <td>675.0</td>\n",
       "      <td>1252.0</td>\n",
       "      <td>1075.0</td>\n",
       "      <td>720.0</td>\n",
       "      <td>784.0</td>\n",
       "      <td>943.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>787.0</td>\n",
       "      <td>983.0</td>\n",
       "      <td>1235.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>174.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>282.0</td>\n",
       "      <td>312.0</td>\n",
       "      <td>303.0</td>\n",
       "      <td>205.0</td>\n",
       "      <td>...</td>\n",
       "      <td>146.0</td>\n",
       "      <td>279.0</td>\n",
       "      <td>262.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>241.0</td>\n",
       "      <td>323.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>184.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>382.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17382 columns</p>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "execution_count": 66
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "source": [
    "full_data_df.shape"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "(31995, 17382)"
      ]
     },
     "metadata": {},
     "execution_count": 67
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Compute filtered signatures"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "source": [
    "from maayanlab_bioinformatics.dge.limma_voom import limma_voom_differential_expression\n",
    "from maayanlab_bioinformatics.normalization.filter import filter_by_expr"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "source": [
    "comparisons = {}\n",
    "for tissue in sample_meta['SMTS'].unique():\n",
    "    sub_meta = sample_meta[sample_meta['SMTS'] == tissue]\n",
    "    data_df = full_data_df[sub_meta.index.tolist()]\n",
    "    data_df = filter_by_expr(data_df)\n",
    "    if not sub_meta[sub_meta['age'] == '20-29'].shape[0] >= 3: \n",
    "        print(tissue, \"not enough healthy samples\")\n",
    "        continue\n",
    "    for agegrp in sub_meta['age'].unique():\n",
    "        if agegrp == '20-29': continue\n",
    "        if sub_meta[sub_meta['age'] == agegrp].shape[0] >= 3:\n",
    "            min_samp = min(sub_meta[sub_meta['age']=='20-29'].shape[0], sub_meta[sub_meta['age']==agegrp].shape[0])\n",
    "            ctl_ids = sub_meta[sub_meta['age'] == '20-29'].sample(n=min_samp, random_state=1).index.tolist()\n",
    "            pert_ids = sub_meta[sub_meta['age'] == agegrp].sample(n=min_samp, random_state=1).index.tolist()\n",
    "            ctl_df = data_df[ctl_ids]\n",
    "            pert_df = data_df[pert_ids]\n",
    "            limma_voom_differential_expression(ctl_df, pert_df).sort_index() \\\n",
    "                .to_csv(f\"GTEx_AgeComparison_Tissue_filtered/GTEx_{tissue.replace(' ', '')}_20-29_vs_{agegrp}.tsv\", sep='\\t')\n",
    "            comparisons[f\"GTEx_{tissue.replace(' ', '')}_20-29_vs_{agegrp}\"] = {'controls': ctl_ids, 'cases': pert_ids}"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "WARNING:root:No group or design set. Assuming all samples belong to one group.\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: R.oo\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: R.methodsS3\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: R.methodsS3 v1.8.1 (2020-08-26 16:20:06 UTC) successfully loaded. See ?R.methodsS3 for help.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: R.oo v1.24.0 (2020-08-26 16:11:58 UTC) successfully loaded. See ?R.oo for help.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: ‘R.oo’\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following object is masked from ‘package:R.methodsS3’:\n",
      "\n",
      "    throw\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:methods’:\n",
      "\n",
      "    getClasses, getMethods\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    attach, detach, load, save\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: R.utils v2.10.1 (2020-08-26 22:50:31 UTC) successfully loaded. See ?R.utils for help.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: ‘R.utils’\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following object is masked from ‘package:utils’:\n",
      "\n",
      "    timestamp\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    cat, commandArgs, getOption, inherits, isOpen, nullfile, parse,\n",
      "    warnings\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: ‘RCurl’\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following object is masked from ‘package:R.utils’:\n",
      "\n",
      "    reset\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following object is masked from ‘package:R.oo’:\n",
      "\n",
      "    clone\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: stats4\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: parallel\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    I, expand.grid, unname\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: IRanges\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following object is masked from ‘package:R.oo’:\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: MatrixGenerics\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: Biobase\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: ‘limma’\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following object is masked from ‘package:DESeq2’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Get metadata"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "source": [
    "names = [cname for cname, _ in comparisons.items()]\n",
    "cases = [cdict['cases'] for _, cdict in comparisons.items()]\n",
    "ctls = [cdict['controls'] for _, cdict in comparisons.items()]\n",
    "\n",
    "pd.DataFrame(\n",
    "    data=[cases, ctls],\n",
    "    index=['pert_samples', 'ctl_samples'],\n",
    "    columns=names\n",
    ").T.to_csv('GTEx_AgeComparison_Tissue_metadata.tsv', sep='\\t')"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [],
   "outputs": [],
   "metadata": {}
  }
 ],
 "metadata": {
  "orig_nbformat": 4,
  "language_info": {
   "name": "python",
   "version": "3.8.10",
   "mimetype": "text/x-python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "pygments_lexer": "ipython3",
   "nbconvert_exporter": "python",
   "file_extension": ".py"
  },
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.10 64-bit ('scatterenv': conda)"
  },
  "interpreter": {
   "hash": "ae02de43a7a1893656bc45d2b0007624c64abadb1f146ae593b6588f2192c0f4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}